lib/internal/errors.js

```diff
@@ -370,42 +370,42 @@ function makeSystemErrorWithCode(key) {
 }
 
 function makeNodeErrorWithCode(Base, key) {
-  return function NodeError(...args) {
-    const limit = Error.stackTraceLimit;
-    if (isErrorStackTraceLimitWritable()) Error.stackTraceLimit = 0;
-    const error = new Base();
-    // Reset the limit and setting the name property.
-    if (isErrorStackTraceLimitWritable()) Error.stackTraceLimit = limit;
-    const message = getMessage(key, args, error);
-    ObjectDefineProperties(error, {
-      [kIsNodeError]: {
-        __proto__: null,
-        value: true,
-        enumerable: false,
-        writable: false,
-        configurable: true,
-      },
-      message: {
-        __proto__: null,
-        value: message,
-        enumerable: false,
-        writable: true,
-        configurable: true,
-      },
-      toString: {
-        __proto__: null,
-        value() {
-          return `${this.name} [${key}]: ${this.message}`;
-        },
-        enumerable: false,
-        writable: true,
-        configurable: true,
-      },
-    });
-    captureLargerStackTrace(error);
-    error.code = key;
-    return error;
-  };
+  const defaultMessage = messages.get(key);
+  const expectedLength = typeof defaultMessage !== 'string' ? -1 : getExpectedArgumentLength(defaultMessage);
+
+  class NodeError extends Base {
+    #message = '';
+    code = key;
+
+    constructor(...args) {
+      super();
+      this.message = expectedLength === 0 && args.length === 0 ?
+        defaultMessage :
+        getMessage(key, args, this);
+    }
+
+    get ['constructor']() {
+      return Base;
+    }
+
+    get [kIsNodeError]() {
+      return true;
+    }
+
+    get message() {
+      return this.#message;
```

Comment:
My guess is, we use `node::PrintCaughtException` to print the error directly during snapshot building, which ultimately invokes `err->ToDetailString()` that is intended to be side-effect-free so the getter would not be invoked.  IMO it's not a good idea for the errors to trigger JS-land getters to get messages anyway, for example if an error is triggered from C++ land, we are not always able to call into JS (the environment could be on the shut down path for some reason, or the environment is not yet fully bootstrapped, or there is an important assertion failed), to make it more robust this should be made a data property.